pkg/vfs/couchdb_indexer.go

```diff
@@ -483,247 +483,138 @@ func (c *couchdbIndexer) setTrashedForFilesInsideDir(doc *DirDoc, trashed bool)
 	return couchdb.BulkUpdateDocs(c.db, consts.Files, files, olddocs)
 }
 
-// TreeFile represent a subset of a file/directory structure that can be used
-// in a tree-like representation of the index.
-type TreeFile struct {
-	DirOrFileDoc
-	FilesChildren     []*TreeFile `json:"children,omitempty"`
-	FilesChildrenSize int64       `json:"children_size,omitempty"`
-	DirsChildren      []*TreeFile `json:"directories,omitempty"`
-
-	isDir    bool
-	hasCycle bool
-	visited  bool
-	errs     []*treeError
-}
-
-// Clone is part of the couchdb.Doc interface
-func (t *TreeFile) Clone() couchdb.Doc {
-	panic("TreeFile must not be cloned")
-}
-
-var _ couchdb.Doc = &TreeFile{}
-
-type treeError struct {
-	file *TreeFile
-	path string
-}
-
-func (c *couchdbIndexer) CheckIndexIntegrity() (logs []*FsckLog, err error) {
-	root, orphans, err := checkIndexIntegrity(func(cb func(entry *TreeFile)) error {
-		return couchdb.ForeachDocs(c.db, consts.Files, func(_ string, data json.RawMessage) error {
-			var f TreeFile
-			if err = json.Unmarshal(data, &f); err == nil {
-				cb(&f)
-			}
-			return err
-		})
-	})
+func (c *couchdbIndexer) CheckIndexIntegrity(predicate func(*FsckLog)) (err error) {
+	root, dirsmap, orphans, err := c.BuildTree()
 	if err != nil {
 		return
 	}
-	if root == nil {
-		logs = append(logs, &FsckLog{Type: IndexMissingRoot})
-		return
+
+	cleanDirsMap(root, dirsmap, predicate)
+
+	for _, entries := range orphans {
+		for _, f := range entries {
+			if f.IsDir {
+				cleanDirsMap(f, dirsmap, predicate)
+			}
+		}
 	}
 
-	logs, err = getLocalTreeLogs(c, root)
-	if err != nil {
-		return
+	for _, orphanCycle := range dirsmap {
+		orphanCycle.hasCycle = true
+		orphans[orphanCycle.DirID] = append(orphans[orphanCycle.DirID], orphanCycle)
 	}
 
-	for dirID, orphansTree := range orphans {
-		for _, entry := range orphansTree {
-			var log *FsckLog
-			log, err = getOrphanTreeLog(c, dirID, entry)
-			if err != nil {
-				return
+	for _, orphansTree := range orphans {
+		for _, orphan := range orphansTree {
+			// TODO: For now, we re-attach the orphan trees at the root of the user's
+			// directory. We may use the dirID to infer more precisely where we
+			// should re-attach this orphan tree. However we should be careful and
+			// check if this directory is actually attached to the root.
+			if !orphan.IsDir {
+				predicate(&FsckLog{
+					Type:    IndexOrphanTree,
+					IsFile:  true,
+					FileDoc: orphan,
+				})
+			} else {
+				predicate(&FsckLog{
+					Type:   IndexOrphanTree,
+					IsFile: false,
+					DirDoc: orphan,
+				})
 			}
-			logs = append(logs, log)
 		}
 	}
 
 	return
 }
 
-func (c *couchdbIndexer) BuildTree() (root *TreeFile, err error) {
-	orphans := make(map[string][]*TreeFile, 32)
-	dirsmap := make(map[string]*TreeFile, 256)
+func (c *couchdbIndexer) BuildTree(eaches ...func(*TreeFile)) (root *TreeFile, dirsmap map[string]*TreeFile, orphans map[string][]*TreeFile, err error) {
```

Comment:
If I take my previous example with 3 folders `/`, `/foo/bar/` and `/foo/bar/baz/`, and 1 file `/foo/bar/baz/qux` (`/foo` is missing), `each` will be called for `/`, `/foo/bar/baz/`, and `/foo/bar/baz/qux` (but nor `/foo/bar/`). Is it normal/expected?

Reply:
I'm a bit surprised, because it means that `Fsck` will say `IndexMissing` for `/foo/` but also for `/foo/bar/` even if we have a document for it.